pkg/provider/ecs/ecs.go

```diff
@@ -104,54 +104,37 @@ func (p *Provider) Init() error {
 	return nil
 }
 
-func (p *Provider) createClient(logger zerolog.Logger) (*awsClient, error) {
-	sess, err := session.NewSessionWithOptions(session.Options{
-		SharedConfigState: session.SharedConfigEnable,
-	})
-	if err != nil {
-		return nil, err
+func (p *Provider) createClient(ctx context.Context, logger zerolog.Logger) (*awsClient, error) {
+	optFns := []func(*config.LoadOptions) error{
+		config.WithLogger(logs.NewAWSWrapper(logger)),
 	}
-
-	ec2meta := ec2metadata.New(sess)
-	if p.Region == "" && ec2meta.Available() {
-		logger.Info().Msg("No region provided, querying instance metadata endpoint...")
-		identity, err := ec2meta.GetInstanceIdentityDocument()
-		if err != nil {
-			return nil, err
-		}
-		p.Region = identity.Region
+	if p.Region != "" {
+		optFns = append(optFns, config.WithRegion(p.Region))
+	} else {
+		logger.Info().Msg("No region provided, will retrieve region from the EC2 Metadata service")
+		optFns = append(optFns, config.WithEC2IMDSRegion())
 	}
 
-	cfg := aws.NewConfig().
-		WithCredentials(credentials.NewChainCredentials([]credentials.Provider{
-			&credentials.StaticProvider{
-				Value: credentials.Value{
-					AccessKeyID:     p.AccessKeyID,
-					SecretAccessKey: p.SecretAccessKey,
-				},
-			},
-			&credentials.EnvProvider{},
-			&credentials.SharedCredentialsProvider{},
-			defaults.RemoteCredProvider(*(defaults.Config()), defaults.Handlers()),
-			stscreds.NewWebIdentityRoleProviderWithOptions(
-				sts.New(sess),
-				os.Getenv("AWS_ROLE_ARN"),
-				"",
-				stscreds.FetchTokenPath(os.Getenv("AWS_WEB_IDENTITY_TOKEN_FILE")),
-			),
-		}))
-
-	// Set the region if it is defined by the user or resolved from the EC2 metadata.
-	if p.Region != "" {
-		cfg.Region = &p.Region
+	if p.AccessKeyID != "" && p.SecretAccessKey != "" {
+		// From https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specify-credentials-programmatically:
+		//   "If you explicitly provide credentials, as in this example, the SDK uses only those credentials."
+		// this makes sure that user-defined credentials always have the highest priority
+		optFns = append(optFns, config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(p.AccessKeyID, p.SecretAccessKey, "")))
```

Comment:
Do you think we should use a cache as described in https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#CredentialsCache?

Reply:
:+1:. Will do this in the next commit.